pkg/storage/minio/saver.go

```diff
@@ -7,28 +7,24 @@ import (
 
 	"github.com/gomods/athens/pkg/errors"
 	"github.com/gomods/athens/pkg/observ"
+	moduploader "github.com/gomods/athens/pkg/storage/module"
 	minio "github.com/minio/minio-go"
 )
 
-func (s *storageImpl) Save(ctx context.Context, module, vsn string, mod []byte, zip io.Reader, info []byte) error {
-	const op errors.Op = "storage.minio.Save"
+func (s *storageImpl) Save(ctx context.Context, module, version string, mod []byte, zip io.Reader, info []byte) error {
+	const op errors.Op = "minio.Save"
 	ctx, span := observ.StartSpan(ctx, op.String())
 	defer span.End()
-	dir := s.versionLocation(module, vsn)
-	modFileName := dir + "/" + "go.mod"
-	zipFileName := dir + "/" + "source.zip"
-	infoFileName := dir + "/" + vsn + ".info"
-	_, err := s.minioClient.PutObject(s.bucketName, modFileName, bytes.NewReader(mod), int64(len(mod)), minio.PutObjectOptions{})
-	if err != nil {
-		return errors.E(op, err)
-	}
-	_, err = s.minioClient.PutObject(s.bucketName, zipFileName, zip, -1, minio.PutObjectOptions{})
-	if err != nil {
-		return errors.E(op, err)
-	}
-	_, err = s.minioClient.PutObject(s.bucketName, infoFileName, bytes.NewReader(info), int64(len(info)), minio.PutObjectOptions{})
+
+	err := moduploader.Upload(ctx, module, version, bytes.NewReader(info), bytes.NewReader(mod), zip, s.upload, s.timeout)
 	if err != nil {
-		return errors.E(op, err)
+		return errors.E(op, err, errors.M(module), errors.V(version))
 	}
 	return nil
 }
+
+func (s *storageImpl) upload(ctx context.Context, path, contentType string, stream io.Reader) error {
+	const op errors.Op = "minio.upload"
+	_, err := s.minioClient.PutObject(s.bucketName, path, stream, -1, minio.PutObjectOptions{})
```

Comment:
@manugupt1 
I _think_ passing `-1` as `objectSize` to the `PutObject` call causes the test to fail. 
The problem here is that if we do not provide the `objectSize` then minio will try to allocate ~600MiB per object upload. In our case we're uploading 3 objects at the same time (mod, info, zip) which means it will need 1.8 GiB. The Windows machine on Azure Pipelines has 3.5 GiB memory but most of it is used by the operating system and mongo. That's why we end up with this nice error `VirtualAlloc of 603979776 bytes failed with errno=1455`.

This raises the question if we should use minio client at all? If we do, then we would need to pass the size from the `gogetFetcher` up to the Saver.
cc: @gomods/maintainers

Reply:
@marpio As an alternative, how about sending bytes & length to the` upload` function from `Saver`?

Reply:
If you are up for it, I can probably send in a proposal PR in the next few days?

Reply:
@marpio I was trying to make the change and now I see what you mean. My proposal would not work

Reply:
@manugut1 doing that in Saver would require us to load the zip into memory which is generally not what we want. A better place would be the goGetFetcher because there we can use `os.Stat` to get the size.

Having said that, I'm not sure minio is a viable option for us because of it's memory requirements. We already removed RDBMS support because of that.
We could try passing the size and measure memory consumption.

Reply:
@marpio If we cannot absolutely do that, then yes.

But I had like to think more about this to get it working and possibly concurrently. The tests on Windows do pass if we do not do fetches concurrently.

I still think there are options. If we still think that we should remove it after I have exhausted my options, then let us see what other maintainers have to say on this. Removing is an option, but let us keep the very end if possible?